### PR TITLE
@ #440 | verify ref before checkout

### PR DESCRIPTION
**Pull request by datphan**
connect #440

### CODE REVIEW

File: lib/teracy-dev/location.rb [hunk 1]
```diff
@@ -32,16 +32,24 @@ def self.git_sync(location, sync_existing)
 
             `git fetch origin`
 
+            current_ref = `git rev-parse --verify HEAD`.strip
+
             if ref
               @@logger.debug("Ref detected, checking out #{ref}")
 
-              `git checkout #{ref}`
+              if !current_ref.include? ref
```
Comment (hoatle) [inline lib/teracy-dev/location.rb]: start_with is better

**Comment by hoatle** [comment]
need to check the origin git repo, if it matches the current git config, then does nothing

```
            `git remote remove origin`

            `git remote add origin #{git}`

            `git fetch origin`
```

`git fetch origin` only when ref or branch changes.

**Review by hoatle — CHANGES_REQUESTED** [review]
need to update more

**Comment by hoatle** [comment]
make sure `$ vagrant status` is clean, the current change still make it very annoying

```
$ vagrant status
From https://github.com/teracyhq-incubator/teracy-dev-core
 * [new branch]      develop    -> origin/develop
 * [new branch]      master     -> origin/master
From https://github.com/teracyhq-incubator/teracy-dev-essential
 * [new branch]      develop    -> origin/develop
From https://github.com/teracyhq-incubator/teracy-dev-k8s
 * [new branch]      develop    -> origin/develop
From https://github.com/kubernetes-incubator/kubespray
 * [new branch]        1439br                 -> origin/1439br
 * [new branch]        ant31-patch-1          -> origin/ant31-patch-1
 * [new branch]        default_true_authtoken -> origin/default_true_authtoken
 * [new branch]        fix_bool_assert        -> origin/fix_bool_assert
 * [new branch]        fix_permissions        -> origin/fix_permissions
 * [new branch]        master                 -> origin/master
 * [new branch]        mattymotest            -> origin/mattymotest
 * [new branch]        numbering_includes     -> origin/numbering_includes
 * [new branch]        preempt_ci_instances   -> origin/preempt_ci_instances
 * [new branch]        preinstall-tasks_num   -> origin/preinstall-tasks_num
 * [new branch]        releasetest            -> origin/releasetest
 * [new branch]        releasetest-1          -> origin/releasetest-1
 * [new branch]        revert-1778-etcd_sec_fixes -> origin/revert-1778-etcd_sec_fixes
 * [new branch]        revert-1809-revert-1795-update_roadmap -> origin/revert-1809-revert-1795-update_roadmap
 * [new branch]        revert-1937-disable-api-insecure-port -> origin/revert-1937-disable-api-insecure-port
 * [new branch]        revert-2345-credentials_upgrade_fix -> origin/revert-2345-credentials_upgrade_fix
 * [new branch]        revert-2380-add-opensuse-support -> origin/revert-2380-add-opensuse-support
 * [new branch]        revert-2639-openstack_tenant_id_fix -> origin/revert-2639-openstack_tenant_id_fix
 * [new branch]        revert-2827-testpr     -> origin/revert-2827-testpr
 * [new branch]        rkt_support            -> origin/rkt_support
 * [new branch]        rpm_specs              -> origin/rpm_specs
 * [new branch]        testpr2                -> origin/testpr2
cat: .git/refs/remotes/origin/v2.6.0: No such file or directory
HEAD is now at 8b3ce6e4... bump upgrade tests to v2.5.0 commit (#3087)
fatal: ambiguous argument 'origin/v2.6.0': unknown revision or path not in the working tree.
Use '--' to separate paths from revisions, like this:
'git <command> [<revision>...] -- [<file>...]'
Current machine states:

k8s-01                    not created (virtualbox)

The environment has not yet been created. Run `vagrant up` to
create the environment. If a machine is not created, only the
default provider will be shown. So if a provider is not listed,
then the machine is not created for that environment.

```

**Comment by datphan** [comment]
fixed:
```
$ vagrant status
Current machine states:

node-01                   running (virtualbox)

The VM is running. To stop this VM, you can run `vagrant halt` to
shut it down forcefully, or you can run `vagrant suspend` to simply
suspend the virtual machine. In either case, to restart it again,
simply run `vagrant up`.
```

**Comment by datphan** [comment]
`git fetch origin` on `branch` condition block is impossible to remove, we had to have it or we will not known what the remote ref is like

**Comment by hoatle** [comment]
on branch/tag checking:
`git fetch origin` only when the current branch is ("develop" or "master") or the current branch is different from the config branch.

Moreover, I got this error for kubespray tag: https://github.com/teracyhq-incubator/teracy-dev-k8s/blob/develop/config_default.yaml#L8

```
$ vagrant status
cat: .git/refs/remotes/origin/v2.6.0: No such file or directory
HEAD is now at 8b3ce6e4... bump upgrade tests to v2.5.0 commit (#3087)
fatal: ambiguous argument 'origin/v2.6.0': unknown revision or path not in the working tree.
Use '--' to separate paths from revisions, like this:
'git <command> [<revision>...] -- [<file>...]'
Current machine states:

k8s-01                    not created (virtualbox)

The environment has not yet been created. Run `vagrant up` to
create the environment. If a machine is not created, only the
default provider will be shown. So if a provider is not listed,
then the machine is not created for that environment.
```

**Comment by hoatle** [comment]
try to test with different extension combinations (more scenarios) so that bugs can reveal.

**Comment by datphan** [comment]
if the current branch is `develop`, and on the remote repo the `develop` branch have updated, we have to use `git fetch origin` to know it is updated.

I'll check the later

**Comment by datphan** [comment]
fixed with tag and branch checking before sync

**Review by hoatle — APPROVED** [review]
LGTM

**Pull request merged**